sound/soc/sof/sof-client.c

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+//
+
+#include <linux/debugfs.h>
+#include <linux/errno.h>
+#include <linux/list.h>
+#include <linux/mutex.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include "sof-client.h"
+#include "sof-priv.h"
+
+static void sof_client_connect(struct ancillary_device *adev)
+{
+	struct sof_client_dev *cdev = ancillary_dev_to_sof_client_dev(adev);
+	struct snd_sof_dev *sdev = cdev->sdev;
+
+	/* add to list of SOF client devices */
+	mutex_lock(&sdev->client_mutex);
+	list_add(&cdev->list, &sdev->client_list);
+	mutex_unlock(&sdev->client_mutex);
+}
+
+static void sof_client_disconnect(struct ancillary_device *adev)
+{
+	struct sof_client_dev *cdev = ancillary_dev_to_sof_client_dev(adev);
+	struct snd_sof_dev *sdev = cdev->sdev;
+
+	/* remove from list of SOF client devices */
+	mutex_lock(&sdev->client_mutex);
+	list_del(&cdev->list);
+	mutex_unlock(&sdev->client_mutex);
+}
+
+static void sof_client_adev_release(struct device *dev)
+{
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+	struct sof_client_dev *cdev = ancillary_dev_to_sof_client_dev(adev);
+
+	ida_simple_remove(cdev->client_ida, adev->id);
+	kfree(cdev);
+}
+
+static struct sof_client_dev *sof_client_dev_alloc(struct snd_sof_dev *sdev, const char *name,
+						     struct ida *client_ida)
+{
+	struct sof_client_dev *cdev;
+	struct ancillary_device *adev;
+	int ret;
+
+	cdev = kzalloc(sizeof(*cdev), GFP_KERNEL);
+	if (!cdev)
+		return ERR_PTR(-ENOMEM);
+
+	cdev->sdev = sdev;
+	cdev->connect = sof_client_connect;
+	cdev->disconnect = sof_client_disconnect;
+	cdev->client_ida = client_ida;
+	adev = &cdev->adev;
+	adev->name = name;
+	adev->dev.parent = sdev->dev;
+	adev->dev.release = sof_client_adev_release;
+
+	adev->id = ida_alloc(client_ida, GFP_KERNEL);
+	if (adev->id < 0) {
+		dev_err(sdev->dev, "error: get IDA idx for ancillary device %s failed\n", name);
+		ret =  adev->id;
+		goto err;
+	}
+
+	ret = ancillary_device_initialize(adev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to initialize client dev %s\n", name);
+		ida_simple_remove(client_ida, adev->id);
+		goto err;
+	}
+
+	return cdev;
+
+err:
+	kfree(cdev);
+	return NULL;
+}
+
+int sof_client_dev_register(struct snd_sof_dev *sdev, const char *name, struct ida *client_ida)
+{
+	struct sof_client_dev *cdev;
+	int ret;
+
+	cdev = sof_client_dev_alloc(sdev, name, client_ida);
+	if (IS_ERR(cdev))
```

Comment:
in the error case, sof_client_dev_alloc() returns NULL, so that test will not work.
Maybe change line 57 to return NULL and test here if (!cdev)?

Reply:
it's still problematic @ranj063 

line 62: return NULL.